### PR TITLE
DPE-1799 Add rotation of mysqlrouter logs

### PR DESCRIPTION
**Pull request by shayancanonical**
## Issue
We do not currently have rotation of mysqlrouter logs

## Solution
- Add logrotate config to the workload container
- Use pebble on the workload container to run a logrotate dispatcher script (which calls `logrotate` with the above config)
- Add integration test

## TODO
Attempt to run logrotate as non-root user

### CODE REVIEW

File: metadata.yaml [hunk 1]
```diff
@@ -43,6 +43,6 @@ resources:
   mysql-router-image:
     type: oci-image
     description: OCI image for mysql-router
-    upstream-source: ghcr.io/canonical/charmed-mysql@sha256:3b6a4a63971acec3b71a0178cd093014a695ddf7c31d91d56ebb110eec6cdbe1
+    upstream-source: ghcr.io/canonical/charmed-mysql@sha256:0f5fe7d7679b1881afde24ecfb9d14a9daade790ec787087aa5d8de1d7b00b21
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply metadata.yaml:46]: yes, it is still 8.0.34: https://github.com/canonical/charmed-mysql-rock/pkgs/container/charmed-mysql/132167093?tag=8.0.34-22.04_edge

File: src/rock.py [hunk 2]
```diff
@@ -110,6 +111,41 @@ def update_mysql_router_service(self, *, enabled: bool, tls: bool = None) -> Non
...
+        """
+        command = "python3 /logrotate_dispatcher.py"
+        startup = (
+            ops.pebble.ServiceStartup.ENABLED.value
+            if enabled
+            else ops.pebble.ServiceStartup.DISABLED.value
+        )
+        layer = ops.pebble.Layer(
+            {
+                "summary": "Logrotate dispatcher layer",
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/rock.py:128]: https://github.com/canonical/pebble#layer-specification indicates it only overrides the service with the same name. i ran a small test to confirm this

File: src/rock.py [hunk 3]
```diff
@@ -110,6 +111,41 @@ def update_mysql_router_service(self, *, enabled: bool, tls: bool = None) -> Non
         else:
             self._container.stop(self._SERVICE_NAME)
 
+    def update_logrotate_dispatcher_service(self, *, enabled: bool) -> None:
+        """Update and restart log rotate dispatcher service.
+
+        Args:
+            enabled: Whether log rotate dispatcher service is enabled
+        """
+        command = "python3 /logrotate_dispatcher.py"
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/rock.py]: addressed in 25aed85927a1c87063c1cee23f7f961ea059c4c9

File: src/kubernetes_logrotate.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""logrotate cron configuration"""
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/kubernetes_logrotate.py]: addressed in 25aed85927a1c87063c1cee23f7f961ea059c4c9

File: scripts/logrotate_dispatcher.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Log rotate event dispatcher."""
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply scripts/logrotate_dispatcher.py]: renamed to `logrotate_executor` in 25aed85927a1c87063c1cee23f7f961ea059c4c9

File: tox.ini [hunk 6]
```diff
@@ -31,7 +32,7 @@ commands_pre =
     # is pending review.
     python -c 'import pathlib; import shutil; import subprocess; git_hash=subprocess.run(["git", "describe", "--always", "--dirty"], capture_output=True, check=True, encoding="utf-8").stdout; file = pathlib.Path("charm_version"); shutil.copy(file, pathlib.Path("charm_version.backup")); version = file.read_text().strip(); file.write_text(f"{version}+{git_hash}")'
 
-    poetry export --only main,charm-libs --output requirements.txt
+    poetry export --only main,charm-libs --output requirements.txt --without-hashes
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tox.ini:36]: addressed in 25aed85927a1c87063c1cee23f7f961ea059c4c9

File: src/kubernetes_logrotate.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+import container
+import logrotate
+
+logger = logging.getLogger(__name__)
+
+SYSTEM_USER = "mysql"
+ROOT_USER = "root"
+
+
+class LogRotate(logrotate.LogRotate):
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/kubernetes_logrotate.py:20]: possibly, but i'd rather it is done as a separate effort. additionally, i dont fully see the benefits of moving to the abstract file as the implementations between vm and k8s differ

File: src/rock.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/rock.py]: not sure what you mean to ask. a summary is a field of the service object, and thus gets overwritten when `replace: override` is specified

File: src/kubernetes_logrotate.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/kubernetes_logrotate.py:20]: agreed, but imo we'll introduce fragmentation of code + the code may be similar, but not identical. additionally, i'd prefer if this was not done as part of this PR (currently, the abstract file is the same for vm and k8s charms)

[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/kubernetes_logrotate.py:17]: addressed in 672badde4c3b621b1aae05d8b620cb89d0279315 - will open corresponding vm PR separately

File: src/rock.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/rock.py]: it is possible that pebble keeps track of the summary of the layers, but it is not exposed to the user with any of the pebble commands. Either way, I removed the layer summary in 86e85169eefece5630e519fadc2749ab100820fe

File: src/kubernetes_logrotate.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/kubernetes_logrotate.py:17]: oops, i missed that, thanks! 86e85169eefece5630e519fadc2749ab100820fe

**Review by taurus-forever — APPROVED** [review]

File: src/kubernetes_logrotate.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+        logger.debug("Copying log rotate executor script to workload container")
+        self._logrotate_executor.write_text(
+            pathlib.Path("scripts/logrotate_executor.py").read_text()
+        )
+        logger.debug("Copied log rotate executor to workload container")
+
+        logger.debug("Starting the logrotate executor service")
+        self._container.update_logrotate_executor_service(enabled=True)
+        logger.debug("Started the logrotate executro service")
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/kubernetes_logrotate.py]: addressed in [577db1a](https://github.com/canonical/mysql-router-k8s-operator/pull/143/commits/577db1a695e7a9f5d45672b6f49f73dd179cb0b3)

File: src/logrotate.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+import jinja2
+
+import container
+
+logger = logging.getLogger(__name__)
+
+SYSTEM_USER = "root"
+
+
+class LogRotate(abc.ABC):
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/logrotate.py]: created an abstract property for system_user in [577db1a](https://github.com/canonical/mysql-router-k8s-operator/pull/143/commits/577db1a695e7a9f5d45672b6f49f73dd179cb0b3)

File: src/kubernetes_logrotate.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/kubernetes_logrotate.py:15]: ah, i misread what you wrote. addressed in [577db1a](https://github.com/canonical/mysql-router-k8s-operator/pull/143/commits/577db1a695e7a9f5d45672b6f49f73dd179cb0b3)

File: src/kubernetes_logrotate.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+class LogRotate(logrotate.LogRotate):
+    """logrotate implementation for k8s"""
+
+    def __init__(self, *, container_: container.Container):
+        super().__init__(container_=container_)
+        self._logrotate_executor = self._container.path("/logrotate_executor.py")
+
+    @property
+    def system_user(self) -> str:
+        return "mysql"
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/kubernetes_logrotate.py]: changed to abstract constant in 2c3aafc47010b717f193cbaf66b91b6b225a68ec

File: src/logrotate.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+    def __init__(self, *, container_: container.Container):
+        self._container = container_
+
+        self._logrotate_config = self._container.path("/etc/logrotate.d/flush_mysqlrouter_logs")
+
+    @property
+    @abc.abstractmethod
+    def system_user(self) -> str:
+        """The system user that mysqlrouter runs as."""
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/logrotate.py]: addressed in 2c3aafc47010b717f193cbaf66b91b6b225a68ec

File: src/logrotate.py [hunk 12]
```diff
@@ -20,16 +20,13 @@
 class LogRotate(abc.ABC):
     """logrotate"""
 
+    _SYSTEM_USER = "mysql"
+
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/logrotate.py]: ah, good catch. [26d0368](https://github.com/canonical/mysql-router-k8s-operator/pull/143/commits/26d0368e7ae3d04e1f519b215c2c9c426f704492)

File: src/logrotate.py [hunk 13]
```diff
@@ -20,16 +20,13 @@
...
     def __init__(self, *, container_: container.Container):
         self._container = container_
 
         self._logrotate_config = self._container.path("/etc/logrotate.d/flush_mysqlrouter_logs")
 
-    @property
-    @abc.abstractmethod
-    def system_user(self) -> str:
-        """The system user that mysqlrouter runs as."""
-
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/logrotate.py]: [26d0368](https://github.com/canonical/mysql-router-k8s-operator/pull/143/commits/26d0368e7ae3d04e1f519b215c2c9c426f704492)

File: src/logrotate.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/logrotate.py]: this might not work, as you may need to remove the abstract property in logrotate.py

**Pull request merged**